fl4health/strategies/feddg_ga_strategy.py

```diff
@@ -0,0 +1,382 @@
+
+        Returns:
+            (Tuple[Optional[float], Dict[str, Scalar]]) A tuple containing the aggregated evaluation loss
+                and the aggregated evaluation metrics.
+        """
+        loss_aggregated, metrics_aggregated = super().aggregate_evaluate(server_round, results, failures)
+
+        self.evaluation_metrics = {}
+        for result in results:
+            cid = result[0].cid
+            self.evaluation_metrics[cid] = result[1].metrics
+            # adding the loss to the metrics
+            val_loss_key = FairnessMetricType.LOSS.value
+            self.evaluation_metrics[cid][val_loss_key] = loss_aggregated  # type: ignore
+
+        # Updating the weights at the end of the training round
+        cids = [result[0].cid for result in results]
+        self.update_weights_by_ga(server_round, cids)
+
+        return loss_aggregated, metrics_aggregated
+
+    def weight_and_aggregate_results(self, results: List[Tuple[ClientProxy, FitRes]]) -> NDArrays:
+        """
+        Aggregate results by weighing them against the adjustment weights and then summing them.
+
+        Args:
+            results: (List[Tuple[ClientProxy, FitRes]]) The clients' fit results.
+
+        Returns:
+            (NDArrays) the weighted and aggregated results.
+        """
+
+        aggregated_results: Optional[NDArrays] = None
+        for result in results:
+            cid = result[0].cid
+
+            # initializing adjustment weights for this client if they don't exist yet
+            if cid not in self.adjustment_weights:
+                self.adjustment_weights[cid] = INITIAL_ADJUSTMENT_WEIGHT
+
+            # apply adjustment weights
+            weighted_client_parameters = parameters_to_ndarrays(result[1].parameters)
+            for i in range(len(weighted_client_parameters)):
+                weighted_client_parameters[i] = weighted_client_parameters[i] * self.adjustment_weights[cid]
+
+            # sum weighted parameters
+            if aggregated_results is None:
+                aggregated_results = weighted_client_parameters
+            else:
+                assert len(weighted_client_parameters) == len(aggregated_results)
+                for i in range(len(weighted_client_parameters)):
+                    aggregated_results[i] = aggregated_results[i] + weighted_client_parameters[i]
+
+        assert aggregated_results is not None
+        return aggregated_results
+
+    def update_weights_by_ga(self, server_round: int, cids: List[str]) -> None:
+        """
+        Update the self.adjustment_weights dictionary by calculating the new weights
+        based on the current server round, fit and evaluation metrics.
+
+        Args:
+            server_round: (int) the current server round.
+            cids: (List[str]) the list of client ids that participated in this round.
+        """
+        value_list = []
```

Comment:
Value list is a bit generic, I think what you're calculating here is the average generalization gap (mu)